### PR TITLE
 [SPARK-36563][SQL] dynamicPartitionOverwrite can direct rename to targetPath instead of partition path one by one when targetPath is empty

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
When target path is empty, we can directly rename stagingDir to targetPath avoid to rename path one by one

### Why are the changes needed?
Optimize file commit logic

### Does this PR introduce _any_ user-facing change?
User can set `spark.sql.source.stagingDir` to enable  direct rename to targetPath instead of partition path one by one when targetPath is empty for dynamicPartitionOverWrite

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Comment by AngersZhuuuu** [comment]
also ping @maropu @dongjoon-hyun @viirya

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @AngersZhuuuu .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1224,6 +1224,16 @@ object SQLConf {
     .stringConf
     .createOptional
 
+  val FILE_STAGING_DIR =
+    buildConf("spark.sql.source.stagingDir")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1228]: This looks orthogonal what your PR aims. Could you spin off this new configuration as a new PR?

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala [hunk 2]
```diff
@@ -203,26 +203,36 @@ class HadoopMapReduceCommitProtocol(
...
-            fs.mkdirs(finalPartPath.getParent)
-          }
-          val stagingPartPath = new Path(stagingDir, part)
-          if (!fs.rename(stagingPartPath, finalPartPath)) {
-            throw new IOException(s"Failed to rename $stagingPartPath to $finalPartPath when " +
+        val targetPath = new Path(path)
+        val pathExisted = fs.exists(targetPath)
+        if (!pathExisted || fs.listStatus(targetPath).isEmpty) {
+          if ((!pathExisted || (pathExisted && fs.delete(targetPath, true))) &&
+            !fs.rename(stagingDir, targetPath)) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:210]: indentation?

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala [hunk 3]
```diff
@@ -203,26 +203,36 @@ class HadoopMapReduceCommitProtocol(
...
-            // FileSystem API spec on rename op says the rename dest(finalPartPath) must have
-            // a parent that exists, otherwise we may get unexpected result on the rename.
-            fs.mkdirs(finalPartPath.getParent)
-          }
-          val stagingPartPath = new Path(stagingDir, part)
-          if (!fs.rename(stagingPartPath, finalPartPath)) {
-            throw new IOException(s"Failed to rename $stagingPartPath to $finalPartPath when " +
+        val targetPath = new Path(path)
+        val pathExisted = fs.exists(targetPath)
+        if (!pathExisted || fs.listStatus(targetPath).isEmpty) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:208]: Do we have a test coverage for both `fs.exist` and `fs.listStatus`?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 4]
```diff
@@ -189,6 +189,33 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
...
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key ->
+        SQLConf.PartitionOverwriteMode.DYNAMIC.toString,
+        SQLConf.FILE_STAGING_DIR.key -> stagingDir.getAbsolutePath) {
+        withTempDir { d =>
+          withTable("t") {
+            sql(
+              s"""
+                 | create table t(c1 int, p1 int) using parquet partitioned by (p1)
+                 | location '${d.getAbsolutePath}'
+            """.stripMargin)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala:204]: indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 5]
```diff
@@ -189,6 +189,33 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
...
+  test("SPARK-36563: dynamic partition overwrite can rename table path when reasonable") {
+    withTempDir { stagingDir =>
+      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key ->
+        SQLConf.PartitionOverwriteMode.DYNAMIC.toString,
+        SQLConf.FILE_STAGING_DIR.key -> stagingDir.getAbsolutePath) {
+        withTempDir { d =>
+          withTable("t") {
+            sql(
+              s"""
+                 | create table t(c1 int, p1 int) using parquet partitioned by (p1)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala:202]: In general, please capitalize SQL keywords like `CREATE TABLE`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 6]
```diff
@@ -189,6 +189,33 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
...
+            """.stripMargin)
+
+            val df = Seq((1, 2), (3, 4)).toDF("c1", "p1")
+            df.write
+              .partitionBy("p1")
+              .mode("overwrite")
+              .saveAsTable("t")
+            checkAnswer(sql("SELECT * FROM t"), df)
+            checkAnswer(sql("SELECT * FROM t WHERE p1 = 2"), Row(1, 2) :: Nil)
+            checkAnswer(sql("SELECT * FROM t WHERE p1 = 4"), Row(3, 4) :: Nil)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala:213]: Does this test case fail before your PR? IIUC, this doesn't provide a test coverage.

Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala:213]: > Does this test case fail before your PR? IIUC, this doesn't provide a test coverage.

Just to show in new way, it can write data as before?

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:208]: Should have

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala:202]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: Done

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1228]: Let me do this first.
